swupd-inspector/archive.go

```diff
@@ -0,0 +1,78 @@
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/bzip2"
+	"compress/gzip"
+	"fmt"
+	"io"
+
+	"github.com/clearlinux/mixer-tools/swupd"
+)
+
+// TODO: Evaluate where to put this, partially duplicate with code in packs.
+
+type compressedTarReader struct {
+	*tar.Reader
+	compressionCloser io.Closer
+}
+
+func (ctr *compressedTarReader) Close() error {
+	if ctr.compressionCloser != nil {
+		return ctr.compressionCloser.Close()
+	}
+	return nil
+}
+
+// Compression algorithms have "magic" bytes in the beginning of the file to identify them.
+var (
+	gzipMagic  = []byte{0x1F, 0x8B}
+	xzMagic    = []byte{0xFD, '7', 'z', 'X', 'Z', 0x00}
+	bzip2Magic = []byte{'B', 'Z', 'h'}
+)
+
+// newCompressedTarReader creates a struct compatible with tar.Reader reading from uncompressed or
+// compressed input.
+func newCompressedTarReader(rs io.ReadSeeker) (*compressedTarReader, error) {
```

Comment:
Seriously, you copied this function, just use the one already written.

Reply:
We don't export those functions / structs, but maybe its time to go ahead and do it (at least in an internal package). I'll fix this duplication.

Reply:
Done.